classes/payment_manager.php

```diff
@@ -0,0 +1,126 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * A manager class used by payment gateway subplugins.
+ *
+ * @package     tool_paymentplugin
+ * @author      Mitchell Halpin
+ *
+ * @copyright   MAHQ
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ **/
+
+namespace tool_paymentplugin;
+
+defined ('MOODLE_INTERNAL') || die();
+
+class payment_manager {
+
+    const CURRENCY = array('USD' => 'usd', 'AUD' => 'aud');
+
+    const PAYMENT_FAILED = 0;
+    const PAYMENT_COMPLETE = 1;
+    const PAYMENT_INCOMPLETE = 2;
+
+    /**
+     * Enrols a user in a course.
+     *
+     * @param int $courseid
+     * @param int $userid
+     * @throws \moodle_exception
+     */
+    public static function paymentplugin_enrol(int $courseid, int $userid) {
+        global $DB;
+        if (!$DB->record_exists('course', array('id' => $courseid))) {
+            throw new \moodle_exception('errorinvalidcourse', 'tool_paymentplugin', '', $courseid);
+        }
+
+        if (!$DB->record_exists('user', array('id' => $userid))) {
+            throw new \moodle_exception('errorinvaliduser', 'tool_paymentplugin', '', $userid);
+        }
+
+        $enrolinstance = $DB->get_record('enrol', array('enrol' => 'payment', 'courseid' => $courseid));
+        if (!$enrolinstance) {
+            throw new \moodle_exception('errorinvalidcourseenrol', 'tool_paymentplugin', '', $courseid);
+        }
+
+        $enrol = enrol_get_plugin('payment');
+        $enrol->enrol_user($enrolinstance, $userid);
+    }
+
+    public static function filter_underscores($data) : \stdClass {
```

Comment:
PHP docblock here. Also, this clones the object, but PHP objects are passed by ref, so it would be slightly more efficient to operate on the passed object, rather than a new object